File: dhrms-backend/src/main/java/com/dhrms/dhrms_backend/controller/DoctorController.java
package com.dhrms.dhrms_backend.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.dhrms.dhrms_backend.dto.doctor.CreateDoctorRequest;
import com.dhrms.dhrms_backend.dto.doctor.DoctorResponse;
import com.dhrms.dhrms_backend.dto.doctor.UpdateDoctorRequest;
import com.dhrms.dhrms_backend.security.AuthenticatedUserService;
import com.dhrms.dhrms_backend.service.DoctorService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/hospitals/doctors")
@RequiredArgsConstructor
@PreAuthorize("hasRole('HOSPITAL')")
public class DoctorController {

        private final AuthenticatedUserService authenticatedUserService;
        private final DoctorService doctorService;

        @PostMapping
        public ResponseEntity<DoctorResponse> createDoctor(
                        Authentication authentication,
                        @Valid @RequestBody CreateDoctorRequest request) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity
                                .status(HttpStatus.CREATED)
                                .body(
                                                doctorService.createDoctor(
                                                                hospitalUserId,
                                                                request));
        }

        @GetMapping
        public ResponseEntity<List<DoctorResponse>> getDoctors(
                        Authentication authentication) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.getHospitalDoctors(
                                                hospitalUserId));
        }

        @GetMapping("/hospital/doctors")
        @PreAuthorize("hasRole('HOSPITAL')")
        public ResponseEntity<List<DoctorResponse>> getHospitalDoctors(
                        Authentication authentication) {

                Long userId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.getDoctorsForHospital(
                                                userId));
        }

        @GetMapping("/{doctorId}")
        public ResponseEntity<DoctorResponse> getDoctor(
                        Authentication authentication,
                        @PathVariable Long doctorId) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.getDoctor(
                                                hospitalUserId,
                                                doctorId));
        }

        @PutMapping("/{doctorId}")
        public ResponseEntity<DoctorResponse> updateDoctor(
                        Authentication authentication,
                        @PathVariable Long doctorId,
                        @Valid @RequestBody UpdateDoctorRequest request) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.updateDoctor(
                                                hospitalUserId,
                                                doctorId,
                                                request));
        }

        @PatchMapping("/{doctorId}/activate")
        public ResponseEntity<DoctorResponse> activateDoctor(
                        Authentication authentication,
                        @PathVariable Long doctorId) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.activateDoctor(
                                                hospitalUserId,
                                                doctorId));
        }

        @PatchMapping("/{doctorId}/suspend")
        public ResponseEntity<DoctorResponse> suspendDoctor(
                        Authentication authentication,
                        @PathVariable Long doctorId) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.suspendDoctor(
                                                hospitalUserId,
                                                doctorId));
        }

        @PatchMapping("/{doctorId}/deactivate")
        public ResponseEntity<DoctorResponse> deactivateDoctor(
                        Authentication authentication,
                        @PathVariable Long doctorId) {

                Long hospitalUserId = authenticatedUserService
                                .getUser(authentication)
                                .getId();

                return ResponseEntity.ok(
                                doctorService.deactivateDoctor(
                                                hospitalUserId,
                                                doctorId));
        }

}

File: dhrms-backend-nestjs/src/doctor/doctor.controller.ts
import { Body, Controller, Get, Param, Patch, Post, Put, Req, UseGuards } from '@nestjs/common';
import { AuthenticatedRequest, JwtAuthGuard } from '../auth/jwt-auth.guard';
import { Roles } from '../auth/roles.decorator';
import { RolesGuard } from '../auth/roles.guard';
import { CreateDoctorDto } from './dto/create-doctor.dto';
import { UpdateDoctorDto } from './dto/update-doctor.dto';
import { DoctorService } from './doctor.service';

@Controller('api/hospitals/doctors')
@UseGuards(JwtAuthGuard, RolesGuard)
@Roles('HOSPITAL')
export class DoctorController {
  constructor(private readonly doctorService: DoctorService) {}

  @Post()
  createDoctor(@Req() req: AuthenticatedRequest, @Body() body: CreateDoctorDto) {
    return this.doctorService.createDoctor(req.user!.id, body);
  }

  @Get()
  getDoctors(@Req() req: AuthenticatedRequest) {
    return this.doctorService.getHospitalDoctors(req.user!.id);
  }

  @Get('/hospital/doctors')
  getHospitalDoctors(@Req() req: AuthenticatedRequest) {
    return this.doctorService.getDoctorsForHospital(req.user!.id);
  }

  @Get('/:doctorId')
  getDoctor(@Req() req: AuthenticatedRequest, @Param('doctorId') doctorId: string) {
    return this.doctorService.getDoctor(req.user!.id, BigInt(doctorId));
  }

  @Put('/:doctorId')
  updateDoctor(@Req() req: AuthenticatedRequest, @Param('doctorId') doctorId: string, @Body() body: UpdateDoctorDto) {
    return this.doctorService.updateDoctor(req.user!.id, BigInt(doctorId), body);
  }

  @Patch('/:doctorId/activate')
  activateDoctor(@Req() req: AuthenticatedRequest, @Param('doctorId') doctorId: string) {
    return this.doctorService.changeStatus(req.user!.id, BigInt(doctorId), 'ACTIVE');
  }

  @Patch('/:doctorId/suspend')
  suspendDoctor(@Req() req: AuthenticatedRequest, @Param('doctorId') doctorId: string) {
    return this.doctorService.changeStatus(req.user!.id, BigInt(doctorId), 'SUSPENDED');
  }

  @Patch('/:doctorId/deactivate')
  deactivateDoctor(@Req() req: AuthenticatedRequest, @Param('doctorId') doctorId: string) {
    return this.doctorService.changeStatus(req.user!.id, BigInt(doctorId), 'INACTIVE');
  }
}
